src/Compilers/Core/MSBuildTask/Utilities.cs

```diff
@@ -171,12 +171,17 @@ internal static Exception GetLocalizedArgumentException(string errorString,
         /// <summary>
         /// Generate the full path to the tool that is deployed with our build tasks.
         /// </summary>
-        internal static string GenerateFullPathToTool(string toolFileName)
+        internal static string GenerateFullPathToTool(string toolFileName, string? toolsDirectory)
         {
             var buildTask = typeof(Utilities).GetTypeInfo().Assembly;
             var assemblyPath = buildTask.Location;
             var assemblyDirectory = Path.GetDirectoryName(assemblyPath)!;
 
+            if (!string.IsNullOrWhiteSpace(toolsDirectory))
+            {
+                return Path.Combine(assemblyDirectory, toolsDirectory, toolFileName);
```

Comment:
This puts a hard requirement that `$(RoslynToolsDirectory)` is a relative path cause if it's a full path then this will throw. I'm not aware of other directory based properties that have this requirement. Think we should consider accepting a full qualified path here and in the case it's not fully qualified use `assemblyDirectory` as the qualifier (as you have now).

Reply:
> This puts a hard requirement that `$(RoslynToolsDirectory)` is a relative path cause if it's a full path then this will throw.

I don't think `Path.Combine` throws. For example, `Path.Combine("C:/a", "/C:/b", "e.f")` results in `"/C:/b\e.f"`.

Reply:
Wow ... that would surprise me. I didn't realize it had that behavior.

Reply:
That's the main difference between Path.Join and Path.Combine I think.